paddle/fluid/pybind/eager_math_op_patch.cc

```diff
@@ -228,17 +232,66 @@ static PyObject* tensor__add__method(TensorObject* self,
       ret = CallScalarFuction(self_tensor, other, "add");
     }
     return ToPyObject(ret);
+  } else if (PyComplex_Check(other_obj)) {
+    if (is_support_complex(self_tensor.dtype()) == false) {
+      eager_gil_scoped_release guard;
+      self_tensor = cast_ad_func(
+          self_tensor, promoteTypes(self_tensor.dtype(), DataType::COMPLEX64));
+    }
   }
 
   // 2. create or get tensor for other_obj
   paddle::Tensor other_tensor;
+
   if (PyCheckTensor(other_obj)) {
-    auto& self_tensor_ref = self->tensor;
-    auto& other_tensor_ref = CastPyArg2Tensor(other_obj, 0);
+    auto& self_tensor_ref_addr = self->tensor;
+    auto& other_tensor_ref_addr = CastPyArg2Tensor(other_obj, 0);
     const phi::distributed::ProcessMesh* mesh = nullptr;
-    if (InputsContainDistTensor(&mesh, self_tensor_ref, other_tensor_ref)) {
-      ConvertAllInputsToDistTensor(mesh, self_tensor_ref, other_tensor_ref);
+    if (InputsContainDistTensor(
+            &mesh, self_tensor_ref_addr, other_tensor_ref_addr)) {
+      ConvertAllInputsToDistTensor(
+          mesh, self_tensor_ref_addr, other_tensor_ref_addr);
+    }
+
+    auto self_tensor_ref = self->tensor;
+    auto other_tensor_ref = CastPyArg2Tensor(other_obj, 0);
+    // got 0-d tensor, and need type promotion. The rules same with Tensor +
+    // Scalar.
+    auto self_tensor_size = self_tensor_ref.shape().size();
+    auto other_tensor_size = other_tensor_ref.shape().size();
+    if ((self_tensor_size == 0 || other_tensor_size == 0) &&
+        self_tensor_ref.dtype() != other_tensor_ref.dtype()) {
+      VLOG(5) << "got 0-d tensor and need to do type promotion, x: "
+              << self_tensor_ref.dtype() << " y: " << other_tensor_ref.dtype();
+      // different major types or both 0-d tensor follow with T+T rule.
+      if (!is_common_dtype_for_scalar(self_tensor_ref.dtype(),
+                                      other_tensor_ref.dtype()) ||
+          (self_tensor_size == 0 && other_tensor_size == 0)) {
+        phi::DataType promote_type = GetPromoteDtype(
+            "add", self_tensor_ref.dtype(), other_tensor_ref.dtype());
+        if (self_tensor_ref.dtype() != promote_type) {
+          eager_gil_scoped_release guard;
+          self_tensor_ref = cast_ad_func(self_tensor_ref, promote_type);
+        }
+        if (other_tensor_ref.dtype() != promote_type) {
+          eager_gil_scoped_release guard;
+          other_tensor_ref = cast_ad_func(other_tensor_ref, promote_type);
+        }
+      } else {
+        // common major types follow with tensor: int32(tensor) + int64(scalar)
+        // = int32
+        if (self_tensor_ref.shape().size() == 0) {
+          eager_gil_scoped_release guard;
+          self_tensor_ref =
+              cast_ad_func(self_tensor_ref, other_tensor_ref.dtype());
+        } else {
+          eager_gil_scoped_release guard;
+          other_tensor_ref =
+              cast_ad_func(other_tensor_ref, self_tensor_ref.dtype());
+        }
+      }
```

Comment:
this section is to check whether there is a 0-dimensional tensor, it's better to write as a common function to reduce duplicate code?

Reply:
This right, I will fix it.